src/libraries/System.Transactions.Local/src/System/Transactions/Oletx/OletxTransactionManager.cs

```diff
@@ -0,0 +1,809 @@
+                                    }
+                                    else
+                                    {
+                                        Environment.FailFast(SR.InternalError);
+                                    }
+
+                                    break;
+                                }
+
+                            case ShimNotificationType.ResourceManagerTmDownNotify:
+                                {
+                                    switch (enlistment2)
+                                    {
+                                        case OletxResourceManager resourceManager:
+                                            resourceManager.TMDown();
+                                            break;
+
+                                        case OletxInternalResourceManager internalResourceManager:
+                                            internalResourceManager.TMDown();
+                                            break;
+
+                                        default:
+                                            Environment.FailFast(SR.InternalError);
+                                            break;
+                                    }
+
+                                    break;
+                                }
+
+                            default:
+                                {
+                                    Environment.FailFast(SR.InternalError);
+                                    break;
+                                }
+                        }
+                    }
+                }
+                finally
+                {
+                    if (holdingNotificationLock)
+                    {
+                        holdingNotificationLock = false;
+                        ProcessingTmDown = false;
+                        Monitor.Exit(ProxyShimFactory);
+                    }
+                }
+            }
+            while (shimNotificationType != ShimNotificationType.None);
+        }
+        finally
+        {
+            if (holdingNotificationLock)
+            {
+                holdingNotificationLock = false;
+                ProcessingTmDown = false;
+                Monitor.Exit(ProxyShimFactory);
+            }
+
+            Thread.EndCriticalRegion();
+        }
+
+        if (etwLog.IsEnabled())
+        {
+            etwLog.MethodExit(TraceSourceType.TraceSourceOleTx, $"{nameof(OletxTransactionManager)}.{nameof(ShimNotificationCallback)}");
+        }
+    }
+
+    internal OletxTransactionManager(string nodeName)
+    {
+        lock (ClassSyncObject)
+        {
+            // If we have not already initialized the shim factory and started the notification
+            // thread, do so now.
+            if (ProxyShimFactory == null)
+            {
+                ProxyShimFactory = new DtcProxyShimFactory(ShimWaitHandle);
+
+                ThreadPool.UnsafeRegisterWaitForSingleObject(
```

Comment:
This is basically a notification loop using an EventWaitHandle to trigger ShimNotificationCallback after pushing a new notification; this is a leftover from when native code was doing the firing. It should be possible to simplify considerably with Channels (or possibly even to handle inline in the same thread), but I'd rather avoid the refactoring here and leave things as they are for now.